File: Python/Algorithms/705.py
class MyHashSet:
    """ Use simple separate chaining. """

    # Use primes for size, sizable intervals.
    psizes = [23, 101, 311, 619, 1249, 2251, 6287, 10639]

    class _Bucket:
        __slots__ = 'value', 'next'

        def __init__(self, value, next_=None):
            self.value = value
            self.next = next_

    def __init__(self, size=0, hasher=hash):
        """ Initialize your data structure here. """
        # Separate chaining doesn't really care about
        # load factor afaik.
        self.load_factor = 0.98
        # Use Python built in hash as hasher.
        self.hash = hasher
        self.size = size if size else self.psizes[0]
        self.length = 0

        # initialize, use least required size initially.
        self._storage = [None for _ in range(self.size)]

    def add(self, key: int) -> None:
        index = self.hash(key) % self.size
        # Is there a bucket here?
        if self._storage[index]:
            b = self._storage[index]
            # go to end of linked structure
            if b.value == key:
                return
            while b.next:
                # don't add key if already here.
                if b.value == key:
                    return
                b = b.next
            b.next = self._Bucket(key)
        # No bucket here, add one.
        else:
            self._storage[index] = self._Bucket(key)
        self.length += 1

        # Resize if we surpass load factor.
        if (self.length / self.size) > self.load_factor:
            self._resize_storage()

    def remove(self, key: int):
        index = self.hash(key) % self.size
        # go through bucket and look for entry with value == key
        b = self._storage[index]
        if not b:
            return
        # head of linked structure contains key
        if b.value == key:
            self._storage[index] = b.next
			self.length -= 1
        # key is somewhere in linked structure
        else:
            while b.next:
                if b.next.value == key:
                    # replace next pointer, decrease length
                    b.next = b.next.next
                    self.length -= 1
                    break
                b = b.next

    def contains(self, key: int) -> bool:
        """
        Returns true if this set contains the specified element
        """
        # go through bucket and look for entry with value == key
        index = self.hash(key) % self.size
        b = self._storage[index]
        if not b:
            return False
        while b:
            if b.value == key:
                return True
            b = b.next
        return False

    def _resize_storage(self):
        """ Resize to next size in primes. """
        # find next size to use
        size, i, psizes = self.size, 0, self.psizes
        while size >= psizes[i]:
            i = i + 1
        size = psizes[i]

        # add everything again.
        _ = MyHashSet(size)
        add = _.add
        for b in self._storage:
            while b:
                add(b.value)
                b = b.next

        # Update
        self._storage = _._storage
        self.size = size
        self.length = _.length


###################################################################

class MyHashSet:
    """ Open addressing for conflict resolution. """

    # Dummy slot used for deletion.
    class _dummy:
        pass

    # Use primes for size, sizable intervals.
    psizes = [23, 101, 311, 619, 1249, 2251, 6287, 10639]
    excludes = {None, _dummy}
    load_factor = .7

    @staticmethod
    def probe_linear(k, size, hasher=hash):
        """ Linear probing."""
        i = 0
        while True:
            yield int(hasher(k) + i) % size
            i = i + 1

    @staticmethod
    def probe_quadratic(k, size, hasher=hash):
        """ Quadratic probing. Yields positions. """
        c, i = 1/2, 0
        while True:
            yield int(hasher(k) + 1/2 * i + 1/2 * (i ** 2)) % size
            i = i + 1

    def __init__(self, size=0, probe=None):
        """ Initialize your data structure here. """
        self.prober = probe if probe else self.probe_linear
        self.length = 0
        self._storage = [None for _ in range(size if size else self.psizes[0])]

    def add(self, key: int) -> None:
        """ Add to the set. If value already exists do nothing. """
        size = len(self._storage)
        excludes = self.excludes
        for slot in self.prober(key, size):
            value = self._storage[slot]
            # if it's here, do nothing.
            if value == key:
                return

            # If we find a dummy or a None, add it.
            if value in excludes:
                self._storage[slot] = key
                self.length += 1
                if (self.length / size) >= self.load_factor:
                    self._resize_storage()
                return

    def remove(self, key: int):
        """ Removal. Using lazy deletion whereby a dummy value fills
        the slot of the value removed.
        """
        for slot in self.prober(key, len(self._storage)):
            value = self._storage[slot]
            # found slot, empty, not here.
            if value is None:
                return

            # Replace with a dummy value.
            if value == key:
                self._storage[slot] = self._dummy
                self.length -= 1
                return

    def contains(self, key: int) -> bool:
        """ Membership testing. """
        dummypos = -1
        for slot in self.prober(key, len(self._storage)):
            # found empty position, set.
            value = self._storage[slot]
            if value is None:
                return False
            if value is self._dummy:
                dummypos = slot

            # found
            if value == key:
                # if we've found a dummypos, swap.
                if dummypos != -1:
                    self._storage[dummypos] = self._storage[slot]
                    self._storage[slot] = None
                return True
        # unreachable
        return False

    def _resize_storage(self):
        """ Resize to next largest size in primes. """
        # find next size to use
        size, i, sizes = len(self._storage), 0, self.psizes
        while size >= sizes[i]:
            i = i + 1
        size = sizes[i]

        # rehash everything and add it again.
        # meh, need to pass probe type again.
        tmpS = MyHashSet(size, self.prober)
        excludes, add = self.excludes, tmpS.add
        for i in self._storage:
            if i not in excludes:
                add(i)

        # Update
        self._storage = tmpS._storage
        self.length = tmpS.length

File: Rust/Algorithms/705.rs
// A very daft idea, in general. Rust doesn't make working with linked structures that easy.
// (to the point where I just gave up on 'remove'). This was mainly done so I can remember
// how frustrating it is and google like a madman.

use std::boxed::Box;
use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};

#[derive(Clone, Debug)]
struct HashLink {
    val: i32,
    next: Option<Box<HashLink>>,
}

struct MyHashSet {
    storage: Vec<Option<Box<HashLink>>>,
    length: usize,
    size: usize,
}

impl MyHashSet {
    fn next_prime(prime: usize) -> Option<usize> {
        for p in &[23, 101, 311, 619, 1249, 2251, 6287, 10639, 17291] {
            if *p as usize > prime {
                return Some(*p as usize);
            }
        }
        None
    }

    fn new() -> Self {
        let size = MyHashSet::next_prime(0)
            .expect("prime argument larger than 10639.");
        MyHashSet {
            storage: vec![None; size],
            length: 0,
            size,
        }
    }

    // Otra initialization, with a specified size.
    fn with_size(size: usize) -> Self {
        MyHashSet {
            storage: vec![None; size],
            length: 0,
            size,
        }
    }

    fn resize_storage(&mut self) {
        let size = MyHashSet::next_prime(self.size)
            .expect("prime argument larger than 10639.");
        let mut _s = MyHashSet::with_size(size);

        for (mut slot) in self.storage.iter() {
            while let Some(link) = slot {
                _s.add(link.val);
                slot = &link.next;
            }
        }

        self.size = size;
        self.length = _s.length;
        self.storage.clear();
        self.storage.append(&mut _s.storage);
    }

    fn index(&mut self, value: i32) -> usize {
        let mut hasher = DefaultHasher::new();
        hasher.write_i32(value);
        hasher.finish() as usize % self.size
    }

    fn add(&mut self, key: i32) {
        let index = self.index(key);
        let opt_slot = &self.storage[index];
        if opt_slot.is_none() {
            self.storage[index] = Some(Box::new(HashLink {
                val: key,
                next: None,
            }));
        } else {
            // Search through links to see if we find val:
            let mut next = opt_slot;
            while let Some(slot) = next {
                if slot.val == key {
                    return;
                }
                if slot.next.is_none() {
                    break;
                }
                next = &slot.next;
                6940166488
            }
            // There's definitely a better way.
            // Can't find it yet.
            let head = Box::new(HashLink {
                val: key,
                next: self.storage[index].take(),
            });
            self.storage[index] = Some(head);
        }
        self.length += 1;
        if ((self.length as f32 / self.size as f32) >= 0.95) {
            self.resize_storage();
        }
    }

    fn remove(&mut self, key: i32) {
        let index = self.index(key);
        let mut slot = self.storage[index].take();
        if slot.is_none() {
            return;
        }
        // I honestly cannot find another way
        // to remove a node from middle of linked
        // structure. Just add everything except for
        // the link val.
        while let Some(link) = slot {
            if link.val != key {
                self.add(link.val);
            }
            // Adjust increments in length by add.
            self.length -= 1;
            slot = link.next;
        }
        self.length -= 1;
    }

    /** Returns true if this set contains the specified element */
    fn contains(&mut self, key: i32) -> bool {
        let index = self.index(key);
        let mut opt_slot = &self.storage[index];
        if opt_slot.is_none() {
            return false;
        }
        while let Some(slot) = opt_slot {
            if slot.val == key {
                return true;
            }
            opt_slot = &slot.next;
        }
        false
    }
}

//////////////////////////////////////////////////////////////////////////
// Open-addressing, a much saner-approach.

use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};

#[derive(Clone, Debug, Copy)]

// Maybe not the best idea to export names with use?
enum Slot {
    Occupied(i32),
    Empty,
    Deleted,
}
use Slot::*;

struct MyHashSet {
    storage: Vec<Slot>,
    length: usize,
}

impl MyHashSet {
    // Constant stuff.
    const load_factor: f32 = 0.7;
    const primes: [usize; 8] =
        [23, 101, 311, 619, 1249, 2251, 6287, 10639];

    // Get the next largest prime after argument prime.
    fn next_prime(prime: usize) -> Option<usize> {
        for p in &MyHashSet::primes {
            if *p > prime {
                return Some(*p);
            }
        }
        None
    }

    // Hash i32 value with DefaultHasher used by HashMap (I guess?)
    fn hash(value: i32) -> usize {
        let mut hasher = DefaultHasher::new();
        hasher.write_i32(value);
        hasher.finish() as usize
    }

    fn new() -> Self {
        // Normally, unwrap_or would use a larger prime, here we know
        // that this bound is sufficient.
        let size = MyHashSet::next_prime(0).unwrap_or(17291);
        MyHashSet {
            storage: vec![Empty; size],
            length: 0,
        }
    }

    // Otra initialization, with a specified size.
    fn with_size(size: usize) -> Self {
        MyHashSet {
            storage: vec![Empty; size],
            length: 0,
        }
    }

    // TODO: Add enum arg to control linear-quadratic?
    fn probe(key: i32, step: usize, size: usize) -> usize {
        (MyHashSet::hash(key) + step) % size
    }

    // Resize storage: done by creating a new temporary HashSet,
    // adding all occupied slots over and then copying the
    // underlying storage.
    fn resize_storage(&mut self) {
        let size =
            MyHashSet::next_prime(self.storage.len()).unwrap_or(17291);
        let mut _s = MyHashSet::with_size(size);

        // Transfer occupied slots over.
        for i in self.storage.iter() {
            if let Occupied(value) = *i {
                _s.add(value);
            }
        }

        // The old switch-a-roo.
        self.length = _s.length;
        self.storage.clear();
        self.storage.append(&mut _s.storage);
    }

    // Add element to hashset, return if already present.
    fn add(&mut self, key: i32) {
        let (mut attempt, size) = (0, self.storage.len());
        let probe = MyHashSet::probe;
        let mut index = probe(key, attempt, size);

        loop {
            match &self.storage[index] {
                Occupied(value) => {
                    // If we're already here, return.
                    if *value == key {
                        return;
                    }
                }
                Empty | Deleted => {
                    // Add if we find empty/deleted slot.
                    self.storage[index] = Occupied(key);
                    self.length += 1;
                    if ((self.length as f32 / size as f32)
                        >= MyHashSet::load_factor)
                    {
                        self.resize_storage();
                    }
                    return;
                }
            }
            // Try next slot.
            attempt += 1;
            index = probe(key, attempt, size);
        }
    }

    // Remove element from set, return if not present.
    fn remove(&mut self, key: i32) {
        let (mut attempt, size) = (0, self.storage.len());
        let probe = MyHashSet::probe;
        let mut index = probe(key, attempt, size);

        loop {
            match &self.storage[index] {
                Empty => {
                    return;
                }
                Deleted => {
                    // Keep going, don't bail.
                }
                Occupied(value) => {
                    // Replace with Slot::Deleted.
                    if *value == key {
                        self.storage[index] = Deleted;
                        self.length -= 1;
                        return;
                    }
                }
            }
            // Try next slot.
            attempt += 1;
            index = probe(key, attempt, size);
        }
    }

    // Returns true if this set contains the specified element.
    fn contains(&mut self, key: i32) -> bool {
        let (mut attempt, mut ilast_deleted) = (0, -1 as i32);
        let size = self.storage.len();
        let probe = MyHashSet::probe;
        let mut index = probe(key, attempt, size);

        loop {
            match &self.storage[index] {
                Occupied(value) => {
                    if *value == key {
                        if ilast_deleted != -1 {
                            // Put at beginning. (Reduce Deleted slots and
                            // find earlier if a search for it is performed again).
                            self.storage[ilast_deleted as usize] =
                                Occupied(key);
                            self.storage[index] = Empty;
                        }
                        return true;
                    }
                }
                Empty => {
                    return false;
                }
                Deleted => {
                    ilast_deleted = index as i32;
                }
            }
            // Try next slot.
            attempt += 1;
            index = probe(key, attempt, size);
        }
        false
    }
}
